### PR TITLE
Added Webapp Manifest Creator Plugin

### PR DESCRIPTION
**Pull request by Symphoria**
I am using 'fs' node package to create the **manifest.json** file.

### CODE REVIEW

File: create-manifest.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module.exports = function createWebappManifest(details) {
+    let manifestObj =  {
+        "name": details.name,
+        "short_name": details.name,
```
Comment (sendilkumarn) [inline create-manifest.js]: If the name is multi word can we shrink it into smaller name 

`The name (maximum of 45 characters)`

`The short_name (maximum of 12 characters recommended) `

```json
  "manifest_version": 2,
  "version": "versionString",
```

can we add these too ?

File: generator.js [hunk 2]
```diff
@@ -21,9 +22,19 @@ module.exports = class WebpackGenerator extends Generator {
 
 		this.options.env.configuration.dev.topScope = [];
 
-		return this.prompt([Confirm('serviceWorker', 'Do you want to add Service Worker?')])
+		return this.prompt([Confirm('serviceWorker', 'Do you want to add Service Worker?'),
+							Input('name', 'What is the name of your application?'),
```
Comment (sendilkumarn) [inline generator.js]: Can specify defaults straightaway here? 

https://github.com/SBoudrias/Inquirer.js/#objects

File: create-manifest.js (hunk 1 above)
Comment (Symphoria) [reply create-manifest.js]: I have exams till Saturday. Would it be alright if I submit the patch incorporating these changes on Saturday?

File: dev-config.js [hunk 3]
```diff
@@ -8,6 +12,9 @@ module.exports = function createDevConfig(serviceWorker, favPath) {
 		plugins.push("new FaviconsWebpackPlugin('"+ favPath +"')");
 	}
 
+	let fd = fs.openSync('manifest.json', 'w');
+	fs.writeFile(fd, createManifest(manifestDetails), 'utf8');
```
Comment (dhruvdutt) [inline dev-config.js]: [`writeFile`](https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback) is an asynchronous call, there might be issues if we write this file at other places without waiting for the callback here.

I believe [`createWriteStream`](https://nodejs.org/api/fs.html#fs_fs_createwritestream_path_options) is synchronous. Can we use it here?

Also, `'utf8'` is default encoding option. We can clean this up.

Comment (sendilkumarn) [reply dev-config.js]: probably `writeFileSync`

File: create-manifest.js (hunk 1 above)
Comment (Symphoria) [reply create-manifest.js]: How to shrink the name? Do I remove whitespaces in between the words or just simply take the first 45 characters of the string?

Comment (sendilkumarn) [reply create-manifest.js]: use `validate` function in `inquirer`

Comment (Symphoria) [reply create-manifest.js]: Got It

**Comment by sendilkumarn** [comment]
can check the failing test cases ?

File: generator.js [hunk 4]
```diff
@@ -30,6 +32,65 @@ module.exports = class WebpackGenerator extends Generator {
...
+					let nameQuestion = {
+						type: 'input',
+						name: 'name',
+						message: "What is the name of your application?",
+						default: () => process.cwd().split(path.sep).pop(),
+						validate: value => {
+							if (value.length <= 45) {
+								return true;
+							} else {
+								return "Name is too long. Please a shorter name (upto 45 characters)";
```
Comment (sendilkumarn) [inline generator.js]: `name is too long, it should be less than 45 characters`

File: create-manifest.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+        "manifest_version": 2,
+        "version": "1.0.0",
+        "start_url": "/" + details.homePage,
+        "display": "standalone",
+        "background_color": details.themeColor,
+        "theme_color": details.themeColor
+    };
+
+    return JSON.stringify(manifestObj);
+};
```
Comment (sendilkumarn) [inline create-manifest.js]: nit: empty line here

**Comment by Symphoria** [comment]
@dhruvdutt @sendilkumarn Commits Squashed into one. The Circe CI check is failing at the `writeFileSync` and `openFileSync` methods in **dev-config.js**. Do I use these methods without *Sync*?

**Comment by dhruvdutt** [comment]
The error is because of this rule.
https://github.com/sendilkumarn/webpack-addons-pwa/blob/d299f9269577f48b86fe37171e5e8472e459cb78/.eslintrc.js#L185

We can either clean it up or use `eslint-disable-line`. @sendilkumarn Thoughts?

**Comment by sendilkumarn** [comment]
@Symphoria are you considering to implement this later.

If manifest is already available?
plugins to handle the manifest file

**Comment by Symphoria** [comment]
Yeah. Will start working on second and third checkboxes in #2

**Comment by Symphoria** [comment]
@sendilkumarn Please review.

File: dev-config.js [hunk 6]
```diff
@@ -1,4 +1,20 @@
-module.exports = function createDevConfig(serviceWorker, favPath) {
+const createManifest = require('./create-manifest');
+const fs = require('fs');
+
+let getManifestPath = manifestDetails => {
+	let manifestPath;
+	if (manifestDetails.hasManifest) {
+		manifestPath = manifestDetails.path;
+	} else {
+		let fd = fs.openSync('manifest.json', 'w');
```
Comment (sendilkumarn) [inline dev-config.js]: can we move this `manifest.json` to a constant field

File: dev-config.js [hunk 7]
```diff
@@ -9,6 +25,10 @@ module.exports = function createDevConfig(serviceWorker, favPath) {
 		plugins.push("new FaviconsWebpackPlugin('" + favPath + "')");
 	}
 
+	if (manifestDetails) {
+		plugins.push("new CopyWebpackPlugin([{ from: '" + getManifestPath(manifestDetails) + "', to: '/build'}])");
```
Comment (sendilkumarn) [inline dev-config.js]: the build should be in the question right. We cannot assume the output dir, so it is better to get this information from users.

File: generator.js [hunk 8]
```diff
@@ -30,6 +33,78 @@ module.exports = class WebpackGenerator extends Generator {
 					);
 				}
 
+				return this.prompt([List('hasManifest', 'Do you have an existing Manifest File?', ['Yes', 'No'])]);
```
Comment (sendilkumarn) [inline generator.js]: can this be a confirm?

**Comment by Symphoria** [comment]
@sendilkumarn Please review.
We were sending too many arguments to `createDevConfig function`, so I made an object and only passed that as an argument. Please do tell if I need to revert that.

**Pull request merged**